### PR TITLE
doc: update lua documentation

### PR DESCRIPTION
**Pull request by teto**
Reflects the recent changes to the lua infrastructure.
Packaging from luarocks should be encouraged.
At first I intended to discard the buildLuaPackage completely (rather alias it to buildLuaRocks) but I guess it makes sense to keep both so I need to update the documentation regarding that.

I believe it might be best to wait a bit to see how the lua infrastructure evolves but I am looking for advice on the documentation (both technical and outline).

I couldn't build it on master:

```
	--out-link ./functions/library/locations.xml
trace: `mkStrict' is obsolete; use `mkOverride 0' instead.
trace: `lib.nixpkgsVersion` is deprecated, use `lib.version` instead!
trace: Warning: `showVal` is deprecated and will be removed in the next release, please use `traceSeqN`
trace: lib.zip is deprecated, use lib.zipAttrsWith instead
/nix/store/x0r93nfw3n7lnb9hpc5xp47qyvr8lgrz-locations.xml
xmllint --nonet --xinclude --noxincludenode manual.xml --output manual-full.xml
warning: failed to load external entity "functions/library/generated/strings.xml"
functions/library.xml:19: element include: XInclude error : could not load functions/library/generated/strings.xml, and no fallback was found
warning: failed to load external entity "functions/library/generated/trivial.xml"
functions/library.xml:20: element include: XInclude error : could not load functions/library/generated/trivial.xml, and no fallback was found
warning: failed to load external entity "functions/library/generated/lists.xml"
functions/library.xml:21: element include: XInclude error : could not load functions/library/generated/lists.xml, and no fallback was found
warning: failed to load external entity "functions/library/generated/debug.xml"
functions/library.xml:22: element include: XInclude error : could not load functions/library/generated/debug.xml, and no fallback was found
warning: failed to load external entity "functions/library/generated/options.xml"
functions/library.xml:23: element include: XInclude error : could not load functions/library/generated/options.xml, and no fallback was found
jing "$RNG" manual-full.xml
/home/teto/nixpkgs3/doc/manual-full.xml:6592:54: error: element "xi:include" not allowed here; expected the element end-tag or element "bibliography", "glossary", "index", "section", "simplesect" or "toc"
/home/teto/nixpkgs3/doc/manual-full.xml:6593:54: error: element "xi:include" not allowed here; expected the element end-tag or element "bibliography", "glossary", "index", "section", "simplesect" or "toc"
/home/teto/nixpkgs3/doc/manual-full.xml:6594:52: error: element "xi:include" not allowed here; expected the element end-tag or element "bibliography", "glossary", "index", "section", "simplesect" or "toc"
/home/teto/nixpkgs3/doc/manual-full.xml:6595:52: error: element "xi:include" not allowed here; expected the element end-tag or element "bibliography", "glossary", "index", "section", "simplesect" or "toc"
/home/teto/nixpkgs3/doc/manual-full.xml:6596:54: error: element "xi:include" not allowed here; expected the element end-tag or element "bibliography", "glossary", "index", "section", "simplesect" or "toc"
make: *** [Makefile:29: validate] Error 1

```

TODO
add lua to the list of setupHooks as well.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: doc/languages-frameworks/lua.section.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+- `withPackages`. Simpler interface to `buildEnv`.
+- `pkgs`. Set of Lua packages for that specific interpreter. The package set can be modified by overriding the interpreter and passing `packageOverrides`.
+
+
+#### `buildLuarocksPackage` function
+
+The `buildLuarocksPackage` function is implemented in `pkgs/development/interpreters/lua-5/build-lua-package.nix`
+The following is an example:
+```nix
+luaexpat = buildLuaPackage rec {
```
Comment (vcunat) [inline doc/languages-frameworks/lua.section.md]: You wanted to use `buildLuarocksPackage` here, I guess?

Comment (teto) [reply doc/languages-frameworks/lua.section.md]: yes. Fixed.

**Comment by teto** [comment]
git blames @grahamc and @vcunat for  lua.xml : so I would like to complete this before the upcoming nixos release. At first, I thought I could get rid of buildLuaPackage along with its doc but now it appears wiser to keep both so I need to merge the old (.xml) and my new doc (.md). How would you recommend to proceed ? I suppose it's best if I merge both the lua.xml and lua.section.md in one file but which one ? I find editing the markdown easier but ultimately am ok with converting markdown to the xml format.

**Comment by vcunat** [comment]
TODO: `s/python/lua/g`

**Comment by teto** [comment]
right also with the "merge" of interpreters, some of the info might be outdated. Will give it another pass if I can build the manual.

**Comment by teto** [comment]
this has been pending for too long. I finally fixed a few mistakes @vcunat if you don't mind having a look.

File: doc/languages-frameworks/lua.section.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+---
+
+# User's Guide to Lua Infrastructure
+
+## Using Lua
+
+### Overview of Lua
+
+Several versions of the Lua interpreter are available: luajit, lua 5.1, 5.2, 5.3.
+The attribute `lua` refers to the default interpreter, it is also possible to refer to specific versions, e.g. `lua_52` refers to Lua 5.2.
```
Comment (lblasc) [inline doc/languages-frameworks/lua.section.md:14]: s/lua_52/lua_5_2/

Comment (teto) [reply doc/languages-frameworks/lua.section.md]: thanks I missed several :(

**Pull request merged**